### PR TITLE
auxdisplay: Enhance SerLCD auxdisplay driver

### PR DESCRIPTION
**Pull request by ShaharHD**
Added export of command and special command delays as configurable options.

Based on the excellent work done by @janhenke, I've done some additional testing on a Nordic board (PCA10056 with nRF52840 chipset) and when trying some high intenesity updates to the LCD, it seems the timeouts might require some tweeking (in contrast to implementing a retry logic in the application)

So I've decided to expose those settings a part of the driver configuration (while leaving the inital values in the code as the defaults)

### CODE REVIEW

**Review by thedjnK — COMMENTED** [review]
Not against this but would prefer it if the defaults were fixed so that they worked in the majority of cases instead. No harm in having too large a timeout but too low a timeout means the display doesn't work, that should be fixed

**Comment by ShaharHD** [comment]
> Not against this but would prefer it if the defaults were fixed so that they worked in the majority of cases instead. No harm in having too large a timeout but too low a timeout means the display doesn't work, that should be fixed

It depends on many factors, but the values set in the current version were copied from SparkFun SerLCD [[SparkFun_SerLCD_Arduino_Library](https://github.com/sparkfun/SparkFun_SerLCD_Arduino_Library)](https://github.com/sparkfun/SparkFun_SerLCD_Arduino_Library/blob/master/src/SerLCD.cpp)

The values seems to be working fine when using I2C at 100 kbps (which is the default for the TWI in many boards).
But when I tried to have a higher refresh rate (400 kpbs) it seems I was hitting a "soft spot" where I had to add some more time as the device itself was still "executing" the command (and resulted in some cases in commmands getting entangled together)

File: drivers/auxdisplay/auxdisplay_serlcd.c [hunk 1]
```diff
@@ -84,6 +74,8 @@ struct auxdisplay_serlcd_data {
 	bool blinking;
 	uint16_t cursor_x;
 	uint16_t cursor_y;
+	int command_delay_ms;
+	int special_command_delay_ms;
 };
```
Comment (nordicjm) [inline drivers/auxdisplay/auxdisplay_serlcd.c:79]: These should be uint, not int, and should be part of the config, not the data

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml [hunk 2]
```diff
@@ -32,3 +34,13 @@ properties:
...
+
+  command-delay:
+    type: int
+    default: 10
+    description: "Delay in milliseconds after a normal command was sent"
+
+  special-command-delay:
+    type: int
+    default: 50
+    description: "Delay in milliseconds after a special command was sent"
```
Comment (nordicjm) [inline dts/bindings/auxdisplay/sparkfun,serlcd.yaml:46]: The quote's are not needed, and defaults need to be justified

**Review by nordicjm — DISMISSED** [review]

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml (hunk 2 above)
Comment (ShaharHD) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml]: removed the quote signs.
but regarding the default values, I kept the exact same indentation as the original code, can you please explain what kind of justification you mean?

image

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 1 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Moved to config.

And changed to uint as requested

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml (hunk 2 above)
Comment (thedjnK) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml]: https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#rules-for-default-values

Comment (ShaharHD) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml]: @thedjnK got it - thanks for the example.

It is better now?
To include the actual link to the SerLCD Sparkfun library?

**Review by thedjnK — DISMISSED** [review]

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml (hunk 2 above)
Comment (thedjnK) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml]: Great, thanks

File: drivers/auxdisplay/auxdisplay_serlcd.c [hunk 3]
```diff
@@ -413,7 +407,8 @@ static const struct auxdisplay_driver_api auxdisplay_serlcd_auxdisplay_api = {
 
 #define AUXDISPLAY_SERLCD_INST(inst)                                                               \
 	static const struct auxdisplay_serlcd_config auxdisplay_serlcd_config_##inst = {           \
-		.capabilities = {                                                                  \
+		.capabilities =                                                                    \
+			{                                                                          \
```
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:411]: the coding style is to have the brace inline, and also a patch should not contain unrelated cosmetic changes, can you revert this bit please? I know clang-format gets confused by this so it has to be handled manually, sorry

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml [hunk 4]
```diff
@@ -32,3 +34,23 @@ properties:
...
+
+  command-delay:
+    type: int
+    default: 10
+    description: |
+      Delay in milliseconds after a normal command was sent.
+      The default value is based on the original SparkFun SerLCD library
+      implementation which assumes 100 kbps I2C configuration. This value
+      might require tweaking if using I2C at a higher bitrare and/or relativily
+      high update frequency of the display.
```
Comment (fabiobaltieri) [inline dts/bindings/auxdisplay/sparkfun,serlcd.yaml:46]: default values should be mentioned in the binding description

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml [hunk 5]
```diff
@@ -32,3 +34,23 @@ properties:
     enum:
       - 2
       - 4
+
+  command-delay:
```
Comment (fabiobaltieri) [inline dts/bindings/auxdisplay/sparkfun,serlcd.yaml:38]: use the unit in the property name, `command-delay-ms`, that way one does not have to cross reference the binding documentation to figure it out

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: I used the .clang-format as documented here:
https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format

```shell
clang-format -i drivers/auxdisplay/auxdisplay_serlcd.c
```

But I actually made it to be **inline** with every other line in the array (`,` at th end of each line)

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml (hunk 4 above)
Comment (ShaharHD) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml:46]: fixed?

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: yeah it's unfortunately common, just make it back to `.capabilities = {` please.

File: dts/bindings/auxdisplay/sparkfun,serlcd.yaml (hunk 4 above)
Comment (fabiobaltieri) [reply dts/bindings/auxdisplay/sparkfun,serlcd.yaml:46]: yup, thanks

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: actually good point, the closing bracket should also be one indentation back

**Review by thedjnK — DISMISSED** [review]

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: you not going to like it :)

```shell
clang-format -i drivers/auxdisplay/auxdisplay_serlcd.c
```

makes all the `\` shift even more right (due to alignment) - because one of the lines got bigger.

where I had it before (matched to **all other** auxdisplay drivers) it was a a 4 lines change, now the whole structure.

File: drivers/auxdisplay/auxdisplay_serlcd.c [hunk 6]
```diff
@@ -413,19 +407,19 @@ static const struct auxdisplay_driver_api auxdisplay_serlcd_auxdisplay_api = {
...
+				 .brightness.minimum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,             \
+				 .brightness.maximum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,             \
+				 .backlight.minimum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,              \
+				 .backlight.maximum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,              \
+				 .custom_characters = SERLCD_CUSTOM_CHAR_MAX_COUNT,                \
+				 .custom_character_width = SERLCD_CUSTOM_CHAR_WIDTH,               \
+				 .custom_character_height = SERLCD_CUSTOM_CHAR_HEIGHT},            \
+		.bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.command_delay_ms = DT_INST_PROP(inst, command_delay_ms),                          \
+		.special_command_delay_ms = DT_INST_PROP(inst, special_command_delay_ms)};         \
```
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:422]: almost :-)

```c
        static const struct auxdisplay_serlcd_config auxdisplay_serlcd_config_##inst = {           \
                .capabilities = {                                                                  \
                        .columns = DT_INST_PROP(inst, columns),                                    \
                        .rows = DT_INST_PROP(inst, rows),                                          \
                        .mode = 0,                                                                 \
                        .brightness.minimum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,                      \
                        .brightness.maximum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,                      \
                        .backlight.minimum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,                       \
                        .backlight.maximum = AUXDISPLAY_LIGHT_NOT_SUPPORTED,                       \
                        .custom_characters = SERLCD_CUSTOM_CHAR_MAX_COUNT,                         \
                        .custom_character_width = SERLCD_CUSTOM_CHAR_WIDTH,                        \
                        .custom_character_height = SERLCD_CUSTOM_CHAR_HEIGHT,                      \
                },                                                                                 \
                .bus = I2C_DT_SPEC_INST_GET(inst),                                                 \
                .command_delay_ms = DT_INST_PROP(inst, command_delay_ms),                          \
                .special_command_delay_ms = DT_INST_PROP(inst, special_command_delay_ms),          \
        };                                                                                         \
```

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: yep :-)

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 6 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: I'm really not seeing what you're seeing.
I reverted my changes, only changed the following:

```c
.bus = I2C_DT_SPEC_INST_GET(inst)};                                                \
```

to

```c
.bus = I2C_DT_SPEC_INST_GET(inst),                                                \
.command_delay_ms = DT_INST_PROP(inst, command_delay_ms), \
.special_command_delay_ms = DT_INST_PROP(inst, special_command_delay_ms)); \
```

And then using [clang-format](https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format) I'm fortmatting before commiting.
I'm not editing manually.

If there's a change to the formatting - this needs to be updated in the official documention.

I'm not manually formatting this.

Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: But this is exacly what I had orginally!

**Review by fabiobaltieri — APPROVED** [review]
Thanks for your patience!

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 6 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: uh, if it is then I messed up, anyway, looks good now, thanks

Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: wait --- it does, the original bracket, hangon

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (fabiobaltieri) [inline drivers/auxdisplay/auxdisplay_serlcd.c:411]: this, restore back to:

```c
		.capabilities = {                                                                  \
```

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 6 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: so many commits and discussions on `{`  ... and I didn't even do it manually but with clang-format ...

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: But that's how .clang-format in zephyr formats it - see all the auxdisplay files - they all look the same

**Comment by thedjnK** [comment]
Just re-request my review once the bracket thing is resolved :)

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: clang-format is known to mess up these definitions, the coding guideline is same as linux: structure instantiations go with the bracket inline with the declaration and field indented one tab. Sorry some reviewers don't check those so there's counter examples in the code, then if your change changes them back in compliance ok, but in this case you are changing it to non compliant AND it's unrelated to your change.

So if you could change it so it's either compliant or as it was before, that would be ok

Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c:411]: > You keep insisting on a change - which all the other files in the same directory are looking different and your clang-format insists on doing the same as the other files.

Yes, I know, I'm sorry, clang-format is known to do changes against the coding style in these cases, I always flag them if I see them.

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 6 above)
Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Yes, I'm asking you to NOT clang-format that block.

**Review by fabiobaltieri — DISMISSED** [review]
Thanks! Sorry for the confusion.

**Comment by fabiobaltieri** [comment]
@thedjnK good to go from the bracket dilemma :-)

**Review by thedjnK — DISMISSED** [review]

File: drivers/auxdisplay/auxdisplay_serlcd.c (hunk 3 above)
Comment (ShaharHD) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: I've manully reverted the line.

Butif not clang-format which one should I use (I'm not keen on manually formatting files...)?

Comment (fabiobaltieri) [reply drivers/auxdisplay/auxdisplay_serlcd.c]: Okay I see. Unfortunately we never got clang-format to fully comply withe coding guidelines for code nested into macros, plus most files are not clang-formatted in the first place, so format on save is not really going to work for upstream work, I'm sorry.

If a file is clang-formatted already you can add `/* clang-format off */` and `/* clang-format on */` on the part where it messes up, and then at least you could rerun clang-format as you wish after that, however, most upstream files are not clang-formatted in the first place and it's normally not accepted to not reformat code unrelated to the change which is why I asked to revert this).

If a file is not clang formatted and you want to clang-format it, it may be ok to do it on a separate commit (as long as you block out cases like the one we discussed here), the maintainer may or may not be happy with that (I would not for the areas I maintain), but I recommend against that because even after you reformatted it, since we don't have any enforcing in CI and since most other contributors indent manually, eventually it's just going to be off again.

**Comment by fabiobaltieri** [comment]
Ok but now you added a merge commit, can you drop it please? We don't use them.

**Comment by ShaharHD** [comment]
> Ok but now you added a merge commit, can you drop it please? We don't use them.

Oh, that was by mistake clicked on the "sync up" on the github.

just realized it...
damn.

**Comment by fabiobaltieri** [comment]
> Oh, that was by mistake clicked on the "sync up" on the github.
> 
> just realized it... damn.

I know it's a trap... odd I thought we got rid of it. Anyway, no prob just reset and force push (or rebase and force push I guess).

**Review by fabiobaltieri — APPROVED** [review]

**Comment by ShaharHD** [comment]
> > Oh, that was by mistake clicked on the "sync up" on the github.
> > just realized it... damn.
> 
> I know it's a trap... odd I thought we got rid of it. Anyway, no prob just reset and force push (or rebase and force push I guess).

So to rebase on latest`main` ?

**Review by thedjnK — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
> So to rebase on latest`main` ? or keep as it is?

It's ok now! Just let the CI finish.
